File: backend/src/services/driver/admin.service.js
const { supabase } = require('../../shared/db/client');
const { AppError } = require('../../shared/middleware/errorHandler');

// ─── Approval Queue ───────────────────────────────────────────────────────────

const getDriverQueue = async ({ status = 'under_review', page = 1, limit = 20 } = {}) => {
  let query = supabase
    .from('drivers')
    .select(`
      id, user_id,
      license_number, license_expiry,
      vehicle_number, vehicle_type, vehicle_make, vehicle_model,
      vehicle_color, vehicle_year,
      approval_status, rejection_reason,
      license_doc_url, vehicle_rc_url, vehicle_insurance_url,
      created_at, updated_at
    `, { count: 'exact' })
    .order('created_at', { ascending: true })
    .range((page - 1) * limit, page * limit - 1);

  if (status !== 'all') {
    query = query.eq('approval_status', status);
  }

  const { data, error, count } = await query;
  if (error) throw new AppError('Failed to fetch driver queue.', 500);

  // Fetch user info separately for each driver (avoids Supabase join ambiguity)
  const drivers = await Promise.all((data || []).map(async (driver) => {
    const { data: user } = await supabase
      .from('users')
      .select('full_name, phone, face_verified, gender')
      .eq('id', driver.user_id)
      .single();
    return { ...driver, users: user || {} };
  }));

  return {
    drivers,
    pagination: {
      page, limit,
      total: count || 0,
      pages: Math.ceil((count || 0) / limit),
    },
  };
};

// ─── Get Single Driver for Review ────────────────────────────────────────────

const getDriverDetail = async (driverId) => {
  const { data, error } = await supabase
    .from('drivers')
    .select(`
      *,
      users!inner(full_name, phone, face_verified, gender, reliability_score, created_at)
    `)
    .eq('id', driverId)
    .single();

  if (error || !data) throw new AppError('Driver not found.', 404);
  return data;
};

// ─── Approve Driver ───────────────────────────────────────────────────────────

const approveDriver = async (driverId, adminUserId) => {
  const driver = await getDriverDetail(driverId);

  if (driver.approval_status === 'approved') throw new AppError('Driver is already approved.', 409);
  if (!driver.users?.face_verified) throw new AppError('Driver must complete face verification before approval.', 403);
  if (driver.license_doc_url === 'pending_upload' || driver.vehicle_rc_url === 'pending_upload') {
    throw new AppError('Driver has not uploaded all required documents.', 400);
  }

  const { error } = await supabase
    .from('drivers')
    .update({
      approval_status: 'approved',
      approved_by: adminUserId,
      approved_at: new Date().toISOString(),
      rejection_reason: null,
    })
    .eq('id', driverId);

  if (error) throw new AppError('Failed to approve driver.', 500);

  console.log(`[Admin] Driver ${driverId} approved by ${adminUserId}`);
  return { approved: true, driverId };
};

// ─── Reject Driver ────────────────────────────────────────────────────────────

const rejectDriver = async (driverId, adminUserId, reason) => {
  if (!reason || reason.trim().length < 10) {
    throw new AppError('Rejection reason must be at least 10 characters.', 400);
  }

  const driver = await getDriverDetail(driverId);
  if (driver.approval_status === 'approved') {
    throw new AppError('Cannot reject an already approved driver. Use suspend instead.', 409);
  }

  const { error } = await supabase
    .from('drivers')
    .update({ approval_status: 'rejected', rejection_reason: reason.trim() })
    .eq('id', driverId);

  if (error) throw new AppError('Failed to reject driver.', 500);

  console.log(`[Admin] Driver ${driverId} rejected by ${adminUserId}: ${reason}`);
  return { rejected: true, driverId, reason };
};

// ─── Suspend Driver ───────────────────────────────────────────────────────────

const suspendDriver = async (driverId, adminUserId, reason) => {
  if (!reason || reason.trim().length < 10) {
    throw new AppError('Suspension reason must be at least 10 characters.', 400);
  }

  const { error } = await supabase
    .from('drivers')
    .update({
      approval_status: 'suspended',
      is_available: false,
      rejection_reason: reason.trim(),
    })
    .eq('id', driverId);

  if (error) throw new AppError('Failed to suspend driver.', 500);

  console.log(`[Admin] Driver ${driverId} suspended by ${adminUserId}: ${reason}`);
  return { suspended: true, driverId, reason };
};

// ─── Reinstate Driver ─────────────────────────────────────────────────────────

const reinstateDriver = async (driverId, adminUserId) => {
  const driver = await getDriverDetail(driverId);
  if (driver.approval_status !== 'suspended') {
    throw new AppError('Driver is not currently suspended.', 409);
  }

  await supabase
    .from('drivers')
    .update({ approval_status: 'approved', rejection_reason: null })
    .eq('id', driverId);

  return { reinstated: true, driverId };
};

// ─── Dashboard Stats ──────────────────────────────────────────────────────────

const getDashboardStats = async () => {
  const { data: drivers } = await supabase
    .from('drivers')
    .select('approval_status, is_available');

  const { data: users } = await supabase
    .from('users')
    .select('role, created_at')
    .eq('is_active', true);

  const weekAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString();

  return {
    drivers: {
      pending:      drivers?.filter(d => d.approval_status === 'pending').length || 0,
      under_review: drivers?.filter(d => d.approval_status === 'under_review').length || 0,
      approved:     drivers?.filter(d => d.approval_status === 'approved').length || 0,
      rejected:     drivers?.filter(d => d.approval_status === 'rejected').length || 0,
      suspended:    drivers?.filter(d => d.approval_status === 'suspended').length || 0,
      online_now:   drivers?.filter(d => d.is_available).length || 0,
    },
    users: {
      passengers:    users?.filter(u => u.role === 'passenger').length || 0,
      drivers:       users?.filter(u => u.role === 'driver').length || 0,
      new_this_week: users?.filter(u => u.created_at >= weekAgo).length || 0,
    },
  };
};

module.exports = {
  getDriverQueue,
  getDriverDetail,
  approveDriver,
  rejectDriver,
  suspendDriver,
  reinstateDriver,
  getDashboardStats,
};

File: mobile/lib/features/admin/admin_service.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/api/api_client.dart';

// ── Data models ───────────────────────────────────────────────────────────────

class DriverQueueItem {
  final String id;
  final String userId;
  final String fullName;
  final String phone;
  final String gender;
  final bool faceVerified;
  final String licenseNumber;
  final String vehicleNumber;
  final String vehicleType;
  final String vehicleMake;
  final String vehicleModel;
  final String vehicleColor;
  final int vehicleYear;
  final String approvalStatus;
  final String? rejectionReason;
  final String licenseDocUrl;
  final String vehicleRcUrl;
  final String? vehicleInsuranceUrl;
  final DateTime createdAt;

  const DriverQueueItem({
    required this.id,
    required this.userId,
    required this.fullName,
    required this.phone,
    required this.gender,
    required this.faceVerified,
    required this.licenseNumber,
    required this.vehicleNumber,
    required this.vehicleType,
    required this.vehicleMake,
    required this.vehicleModel,
    required this.vehicleColor,
    required this.vehicleYear,
    required this.approvalStatus,
    this.rejectionReason,
    required this.licenseDocUrl,
    required this.vehicleRcUrl,
    this.vehicleInsuranceUrl,
    required this.createdAt,
  });

  factory DriverQueueItem.fromJson(Map<String, dynamic> json) {
    final user = json['users'] as Map<String, dynamic>? ?? {};
    return DriverQueueItem(
      id: json['id'] as String? ?? '',
      userId: json['user_id'] as String? ?? '',
      fullName: user['full_name'] as String? ?? 'Unknown',
      phone: user['phone'] as String? ?? '',
      gender: user['gender'] as String? ?? '',
      faceVerified: user['face_verified'] as bool? ?? false,
      licenseNumber: json['license_number'] as String? ?? '',
      vehicleNumber: json['vehicle_number'] as String? ?? '',
      vehicleType: json['vehicle_type'] as String? ?? '',
      vehicleMake: json['vehicle_make'] as String? ?? '',
      vehicleModel: json['vehicle_model'] as String? ?? '',
      vehicleColor: json['vehicle_color'] as String? ?? '',
      vehicleYear: (json['vehicle_year'] as num?)?.toInt() ?? 0,
      approvalStatus: json['approval_status'] as String? ?? 'pending',
      rejectionReason: json['rejection_reason'] as String?,
      licenseDocUrl: json['license_doc_url'] as String? ?? '',
      vehicleRcUrl: json['vehicle_rc_url'] as String? ?? '',
      vehicleInsuranceUrl: json['vehicle_insurance_url'] as String?,
      createdAt: DateTime.tryParse(
              json['created_at'] as String? ?? '') ??
          DateTime.now(),
    );
  }

  String get vehicleDisplay => '$vehicleColor $vehicleMake $vehicleModel ($vehicleYear)';

  String get statusLabel {
    switch (approvalStatus) {
      case 'under_review': return 'Under Review';
      case 'approved':     return 'Approved';
      case 'rejected':     return 'Rejected';
      case 'suspended':    return 'Suspended';
      default:             return 'Pending';
    }
  }
}

class AdminStats {
  final int pending;
  final int underReview;
  final int approved;
  final int onlineNow;
  final int totalPassengers;
  final int newThisWeek;

  const AdminStats({
    required this.pending,
    required this.underReview,
    required this.approved,
    required this.onlineNow,
    required this.totalPassengers,
    required this.newThisWeek,
  });

  factory AdminStats.fromJson(Map<String, dynamic> json) {
    final drivers = json['drivers'] as Map<String, dynamic>? ?? {};
    final users   = json['users']   as Map<String, dynamic>? ?? {};
    return AdminStats(
      pending:        (drivers['pending'] as num?)?.toInt() ?? 0,
      underReview:    (drivers['under_review'] as num?)?.toInt() ?? 0,
      approved:       (drivers['approved'] as num?)?.toInt() ?? 0,
      onlineNow:      (drivers['online_now'] as num?)?.toInt() ?? 0,
      totalPassengers:(users['passengers'] as num?)?.toInt() ?? 0,
      newThisWeek:    (users['new_this_week'] as num?)?.toInt() ?? 0,
    );
  }
}

// ── AdminService ──────────────────────────────────────────────────────────────

class AdminService {
  final ApiClient _api;
  const AdminService(this._api);

  Future<List<DriverQueueItem>> getQueue({
    String status = 'under_review',
    int page = 1,
  }) async {
    final data = await _api.get(
      '/drivers/admin/queue',
      queryParams: {'status': status, 'page': page},
    );
    final list = data['drivers'] as List<dynamic>? ?? [];
    return list
        .map((e) => DriverQueueItem.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<AdminStats> getStats() async {
    final data = await _api.get('/drivers/admin/stats');
    return AdminStats.fromJson(data);
  }

  Future<void> approveDriver(String driverId) async {
    await _api.post('/drivers/admin/$driverId/approve');
  }

  Future<void> rejectDriver(String driverId, String reason) async {
    await _api.post(
      '/drivers/admin/$driverId/reject',
      data: {'reason': reason},
    );
  }

  Future<void> suspendDriver(String driverId, String reason) async {
    await _api.post(
      '/drivers/admin/$driverId/suspend',
      data: {'reason': reason},
    );
  }
}

final adminServiceProvider = Provider<AdminService>((ref) {
  return AdminService(ref.watch(apiClientProvider));
});
